### PR TITLE
Make bootstrap server more robust and customizable.

### PR DESCRIPTION
**Pull request by sbernard31**
1. Add more callback on BootstrapSessionManager for better monitoring.
2. Add a way to customize bootstrapHandler behavior on error.  
3. Improve BootstrapHandler to make it able to avoid to start 2 bootstrap session at the same time for the same device.  
4. Logs enhancement

### CODE REVIEW

**Comment by sbernard31** [comment]
In `DefaultBootstrapHandler`, I asked myself if the executor is really needed ? I would go for remove it. :thinking: 

WDYT ?

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapSessionManager.java [hunk 1]
```diff
@@ -42,6 +43,29 @@
      */
     public BootstrapSession begin(String endpoint, Identity clientIdentity);
 
+    /**
+     * Called when we receive a successful response to a request.
+     * 
+     * @param request The request for which we get a successfull response.
+     */
+    public void onSuccess(LwM2mRequest<? extends LwM2mResponse> request);
```
Comment (msangoi) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapSessionManager.java]: I think the method names can be confusing. e.g. onSuccess() is not about the success of the bootstrap sequence. Or onFailure() and the exisiting failed() are pretty close. 
Maybe onResponseSuccess(), onResponseError(), onRequestFailure()?

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapSessionManager.java [hunk 2]
```diff
@@ -32,6 +31,29 @@
...
+         */
+        RETRYALL,
+        /**
+         * Stop to try to write bootstrap config and directly send a "Bootstrap Finished" request.
+         */
+        SEND_FINISHED,
+        /**
+         * Stop the bootstrap session. The session is considered as failure.
+         */
+        STOP
```
Comment (msangoi) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapSessionManager.java]: FAIL?

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapSessionManager.java [hunk 3]
```diff
@@ -32,6 +31,29 @@
  */
 public interface BootstrapSessionManager {
 
+    public enum BootstrapAction {
```
Comment (msangoi) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapSessionManager.java]: Not really an "Action". But I cannot propose a better name for now 🙂

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/DefaultBootstrapHandler.java [hunk 4]
```diff
@@ -236,21 +234,19 @@ protected void writeSecurities(final BootstrapSession session, final BootstrapCo
...
                         afterWriteSecurities(session, cfg, securityInstancesToWrite, action);
                     }
                 }
             }, new ErrorCallback() {
                 @Override
                 public void onError(Exception e) {
-                    LOG.debug(String.format(
-                            "Error during bootstrap write of security instance %s to foreign peer [ep=%s,id=%s]",
-                            securityInstance, session.getEndpoint(), session.getIdentity()), e);
+                    LOG.debug("Error for {} while sending {} ", session, writeBootstrapRequest, e);
```
Comment (msangoi) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/DefaultBootstrapHandler.java]: I think the exception parameter is going to be ignored here. e will be considered as an argument to the string format.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/DefaultBootstrapHandler.java [hunk 5]
```diff
@@ -135,7 +135,12 @@ public BootstrapResponse bootstrap(Identity sender, BootstrapRequest request) {
         e.execute(new Runnable() {
             @Override
             public void run() {
-                startBootstrap(session, cfg);
+                try {
+                    startBootstrap(session, cfg);
+                } catch (RuntimeException e) {
+                    LOG.warn("Unexpected error at bootstrap start-up for {}", session, e);
```
Comment (msangoi) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/DefaultBootstrapHandler.java]: Same comment about exception as parameter.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapSessionManager.java (hunk 1 above)
Comment (sbernard31) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapSessionManager.java]: :+1:

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapSessionManager.java (hunk 2 above)
Comment (sbernard31) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapSessionManager.java]: :thinking:  ok.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapSessionManager.java (hunk 3 above)
Comment (sbernard31) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapSessionManager.java]: You have all the week-end to think about that. :wink:

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/DefaultBootstrapHandler.java (hunk 4 above)
Comment (sbernard31) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/DefaultBootstrapHandler.java]: I pretty sure it's OK, but I will test this.

Comment (sbernard31) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/DefaultBootstrapHandler.java]:
```java
public static void main(String[] args) {
	LOG.error("My log with params {} {}","param1","param2", new NullPointerException());
}
```
gives
```
17:18:26.659 [main] ERROR test.LOGGER - My log with params param1 param2
java.lang.NullPointerException: null
	at test.LOGGER.main(LOGGER.java:11)
```

So it looks OK

Comment (msangoi) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/DefaultBootstrapHandler.java]: ok 🤔 I was sure I've had this issue in the past. anyway, thanks for testing it.

Comment (sbernard31) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/DefaultBootstrapHandler.java]: Maybe it depends on the slf4j backend :thinking:
I only tested with logback.

Comment (sbernard31) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/DefaultBootstrapHandler.java]: (it's OK for slf4j-simple too)

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapSessionManager.java (hunk 3 above)
Comment (sbernard31) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapSessionManager.java]: Did you find a better name during your week-end :grin:  ?

**Comment by sbernard31** [comment]
Any thought about the executor ?

**Pull request merged**